File: leader-election/src/main/java/com/iluwatar/leaderelection/AbstractInstance.java
/*
 * This project is licensed under the MIT license. 
 * Module model-view-viewmodel is using ZK framework licensed under LGPL (see lgpl-3.0.txt).
 *
 * The MIT License
 * Copyright © 2014-2022 Ilkka Seppälä
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.iluwatar.leaderelection;

import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;
import lombok.extern.slf4j.Slf4j;

/**
 * Abstract base class for all instance implementations in the leader election system.
 * 
 * Each instance runs on its own thread and processes incoming messages asynchronously.
 * This version fixes the busy loop problem by adding a short sleep when no messages are present.
 */
@Slf4j
public abstract class AbstractInstance implements Instance, Runnable {

  // Interval between heartbeats in milliseconds.
  protected static final int HEARTBEAT_INTERVAL = 5000;
  private static final String INSTANCE = "Instance ";

  protected MessageManager messageManager;
  protected Queue<Message> messageQueue;
  protected final int localId;
  protected int leaderId;
  protected boolean alive;

  /**
   * Constructor initializing the instance.
   *
   * @param messageManager manager to send/receive messages
   * @param localId ID of this instance
   * @param leaderId current leader ID
   */
  public AbstractInstance(MessageManager messageManager, int localId, int leaderId) {
    this.messageManager = messageManager;
    this.messageQueue = new ConcurrentLinkedQueue<>();
    this.localId = localId;
    this.leaderId = leaderId;
    this.alive = true;
  }

  /**
   * Thread run loop — continuously processes messages while instance is alive.
   * 
   * 🟢 FIXED: Added small sleep when queue is empty to avoid busy looping.
   */
  @Override
  public void run() {
    while (alive) {
      if (!this.messageQueue.isEmpty()) {
        processMessage(this.messageQueue.poll());
      } else {
        try {
          Thread.sleep(100); // 🔸 Prevents busy loop CPU overuse
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
          LOGGER.warn(INSTANCE + localId + " thread interrupted.");
          break;
        }
      }
    }
    LOGGER.info(INSTANCE + localId + " stopped running.");
  }

  /**
   * Add a new message to the queue to be processed.
   *
   * @param message Message sent by other instances
   */
  @Override
  public void onMessage(Message message) {
    messageQueue.offer(message);
  }

  /** Check if this instance is alive. */
  @Override
  public boolean isAlive() {
    return alive;
  }

  /** Update the alive status of this instance. */
  @Override
  public void setAlive(boolean alive) {
    this.alive = alive;
  }

  /**
   * Process the given message according to its type.
   *
   * @param message message to process
   */
  private void processMessage(Message message) {
    switch (message.getType()) {
      case ELECTION -> {
        LOGGER.info("{}{} - Handling Election Message...", INSTANCE, localId);
        handleElectionMessage(message);
      }
      case LEADER -> {
        LOGGER.info("{}{} - Handling Leader Message...", INSTANCE, localId);
        handleLeaderMessage(message);
      }
      case HEARTBEAT -> {
        LOGGER.info("{}{} - Handling Heartbeat Message...", INSTANCE, localId);
        handleHeartbeatMessage(message);
      }
      case ELECTION_INVOKE -> {
        LOGGER.info("{}{} - Handling Election Invoke...", INSTANCE, localId);
        handleElectionInvokeMessage();
      }
      case LEADER_INVOKE -> {
        LOGGER.info("{}{} - Handling Leader Invoke...", INSTANCE, localId);
        handleLeaderInvokeMessage();
      }
      case HEARTBEAT_INVOKE -> {
        LOGGER.info("{}{} - Handling Heartbeat Invoke...", INSTANCE, localId);
        handleHeartbeatInvokeMessage();
      }
      default -> LOGGER.warn("{}{} - Unknown message type received.", INSTANCE, localId);
    }
  }

  // Abstract methods for handling various message types — to be implemented by subclasses.
  protected abstract void handleElectionMessage(Message message);
  protected abstract void handleElectionInvokeMessage();
  protected abstract void handleLeaderMessage(Message message);
  protected abstract void handleLeaderInvokeMessage();
  protected abstract void handleHeartbeatMessage(Message message);
  protected abstract void handleHeartbeatInvokeMessage();
}

File: leader-election/src/main/kotlin/com/iluwatar/leaderelection/AbstractInstance.kt
/*
 * This project is licensed under the MIT license. Module model-view-viewmodel is using ZK framework licensed under LGPL (see lgpl-3.0.txt).
 *
 * The MIT License
 * Copyright © 2014-2022 Ilkka Seppälä
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.iluwatar.leaderelection

import io.github.oshai.kotlinlogging.KotlinLogging
import java.util.Queue
import java.util.concurrent.ConcurrentLinkedQueue

// ABOUTME: Abstract base class for all instance implementations in the leader election system.
// ABOUTME: Provides message queue handling and dispatches messages to type-specific handlers.

private val logger = KotlinLogging.logger {}

/**
 * Abstract class of all the instance implementation classes.
 */
abstract class AbstractInstance(
    internal var messageManager: MessageManager?,
    internal val localId: Int,
    internal var leaderId: Int
) : Instance, Runnable {

    internal val messageQueue: Queue<Message> = ConcurrentLinkedQueue()
    internal var alive: Boolean = true

    /**
     * The instance will execute the message in its message queue periodically once it is alive.
     */
    @Suppress("squid:S2189")
    override fun run() {
        while (true) {
            if (messageQueue.isNotEmpty()) {
                processMessage(messageQueue.remove())
            }
        }
    }

    /**
     * Once messages are sent to the certain instance, it will firstly be added to the queue and wait
     * to be executed.
     *
     * @param message Message sent by other instances
     */
    override fun onMessage(message: Message) {
        messageQueue.offer(message)
    }

    /**
     * Check if the instance is alive or not.
     *
     * @return `true` if the instance is alive.
     */
    override fun isAlive(): Boolean = alive

    /**
     * Set the health status of the certain instance.
     *
     * @param alive `true` for alive.
     */
    override fun setAlive(alive: Boolean) {
        this.alive = alive
    }

    /**
     * Process the message according to its type.
     *
     * @param message Message polled from queue.
     */
    private fun processMessage(message: Message) {
        when (message.type) {
            MessageType.ELECTION -> {
                logger.info { "$INSTANCE$localId - Election Message handling..." }
                handleElectionMessage(message)
            }
            MessageType.LEADER -> {
                logger.info { "$INSTANCE$localId - Leader Message handling..." }
                handleLeaderMessage(message)
            }
            MessageType.HEARTBEAT -> {
                logger.info { "$INSTANCE$localId - Heartbeat Message handling..." }
                handleHeartbeatMessage(message)
            }
            MessageType.ELECTION_INVOKE -> {
                logger.info { "$INSTANCE$localId - Election Invoke Message handling..." }
                handleElectionInvokeMessage()
            }
            MessageType.LEADER_INVOKE -> {
                logger.info { "$INSTANCE$localId - Leader Invoke Message handling..." }
                handleLeaderInvokeMessage()
            }
            MessageType.HEARTBEAT_INVOKE -> {
                logger.info { "$INSTANCE$localId - Heartbeat Invoke Message handling..." }
                handleHeartbeatInvokeMessage()
            }
            else -> {}
        }
    }

    /**
     * Abstract methods to handle different types of message. These methods need to be implemented in
     * concrete instance class to implement corresponding leader-selection pattern.
     */
    internal abstract fun handleElectionMessage(message: Message)

    internal abstract fun handleElectionInvokeMessage()

    internal abstract fun handleLeaderMessage(message: Message)

    internal abstract fun handleLeaderInvokeMessage()

    internal abstract fun handleHeartbeatMessage(message: Message)

    internal abstract fun handleHeartbeatInvokeMessage()

    companion object {
        internal const val HEARTBEAT_INTERVAL = 5000
        private const val INSTANCE = "Instance "
    }
}
